[GeoMechanicsApplication] Implement Mohr-Coulomb plastic law for continuum elements

mnabideltares opened the pull request:
**📝 Description**
This is to implement Mohr-Coulomb law for plastic behaviour.

For more details:

https://github.com/orgs/Deltares/projects/55/views/11?pane=issue&itemId=95624718&issue=Deltares%7CFEA-Tools%7C146

**🆕 Changelog**
- Added class for Coulomb yield function 
- Added class for tensile cutoff
- Added Mohr-Coulomb class
- Added unittests

mnabideltares committed:
Registered Mohr-Coulomb law

mnabideltares committed:
Added input material variables + defined yield functions + calculated principal stresses

mnabideltares committed:
updated the unit test, fixed a bug in principal stresses

mnabideltares committed:
Added another unit test, modified the variables in the class

mnabideltares committed:
modifications

mnabideltares committed:
restructured the current implementation

mnabideltares committed:
a fix in the unittest

mnabideltares committed:
cleaning

mnabideltares committed:
fix

mnabideltares committed:
Added FlowFunctionDerivateive

mnabideltares committed:
Merge branch 'master' into geo/13071-implement-mohr-coulomb-continuum

mnabideltares committed:
Added PK2

avdg81 committed:
Disabled registration of the Mohr-Coulomb law

markelov208 committed:
added unit tests for Clone and Check

markelov208 committed:
adde unit test for SetValueAndGetValue

markelov208 committed:
removed -// ==== lines

markelov208 committed:
fixed variable name

markelov208 committed:
added more unit tests (some of them not active yet)

mnabideltares committed:
Added trial stresses and reagion detections. Still ongoing process

mnabideltares committed:
fix

markelov208 committed:
Merge branch 'master' into geo/13071-implement-mohr-coulomb-continuum

markelov208 committed:
fixed unit tests

mnabideltares committed:
Added functionalities,

mnabideltares committed:
Implementation and refactoring

mnabideltares committed:
refactoring

mnabideltares committed:
Refactoring and added unit tests

mnabideltares committed:
testing

mnabideltares committed:
unit testing

mnabideltares committed:
refactoring

mnabideltares committed:
refactoring

mnabideltares committed:
Revert "refactoring"

This reverts commit 68a7abf2dbf88175360d02196f2dbd9dd7fd4f51.

mnabideltares committed:
Revert "refactoring"

This reverts commit ac8deee9d44e5d152237216e9bf0310487d8c432.

mnabideltares committed:
refactoring 1

mnabideltares committed:
refactoring 2

mnabideltares committed:
refactor 3

mnabideltares committed:
refactoring and unit testing

mnabideltares committed:
refactoring, and clang format

mnabideltares committed:
code smells

mnabideltares committed:
A fix

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_yield_function.hpp:
```diff
@@ -0,0 +1,42 @@
+private:
+    // Member Variables
+    double mFrictionAngle   = 0.0;
+    double mCohesion        = 0.0;
```
I do not think that storing material properties as member vars is the best option from the memory point of view... Why not using th value from the `Properties`? Member variables should be used for historical kind of information in CLs I think

avdg81 replied:
Note that this class is not a constitutive law, but a yield surface. We feel that storing the parameters that define the yield surface outside of the class makes it more difficult to reason about it locally.

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_constitutive/evaluate_yield_function.h:
```diff
@@ -0,0 +1,29 @@
+    virtual ~EvaluateYieldFunction() = default;
+
+    virtual double CalculateYieldFunction(const Vector& rPrincipalStress) const        = 0;
+    virtual Vector CalculateFlowFunctionDerivate(const Vector& rPrincipalStress) const = 0;
```
I'd add the properties here and the full stress vector for generality... Then each yield will return its own measure of the stress equivalent

avdg81 replied:
This is our interface class, and interface classes are not supposed to own any data. It should only prescribe the way you can interact with concrete classes that derive from this interface.

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_constitutive_law.cpp:
```diff
@@ -0,0 +1,290 @@
+                                      const ProcessInfo&  rCurrentProcessInfo) const
+{
+    // Verify Properties variables
+    if (!rMaterialProperties.Has(GEO_COHESION) || rMaterialProperties[GEO_COHESION] <= 0.0)
```
In the CORE we do have `COHESION` variable I think

avdg81 replied:
The variable `COHESION` is defined by several applications (ConstitutiveLawsApplication, MPMApplication, PfemFluidDynamicsApplication, and PoromechanicsApplication), but not in Kratos Core. We don't want to depend on any of these applications. Furthermore, to avoid name clashes, we have decided to introduce a new variable prefixed with `GEO_` to make its scope/context clear.

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_constitutive_law.cpp:
```diff
@@ -0,0 +1,290 @@
+{
+    Vector deltaStrainVector = rValues.GetStrainVector() - mStrainVectorFinalized;
+    Matrix elastic_matrix    = this->CalculateElasticMatrix(rValues);
+    rStressVector            = mStressVectorFinalized + prod(elastic_matrix, deltaStrainVector);
```
noalias

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_constitutive_law.cpp:
```diff
@@ -0,0 +1,290 @@
+    const auto        E                     = r_material_properties[YOUNG_MODULUS];
+    const auto        NU                    = r_material_properties[POISSON_RATIO];
+
+    const double c0 = E / ((1.0 + NU) * (1.0 - 2.0 * NU));
```
This is implemented in the constitutive_law_utilities

avdg81 replied:
For plane strain models, we use 4x4 matrices, whereas the StructuralMechanicsApplication uses 3x3 matrices.

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_constitutive_law.hpp:
```diff
@@ -0,0 +1,107 @@
+//
+//  Main authors:    Mohamed Nabi,
+//                   Wijtze Pieter Kikstra
+// //
```
double comment

avdg81 replied:
Thanks, we have fixed it.

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_constitutive_law.hpp:
```diff
@@ -0,0 +1,107 @@
+        void SetValue(const Variable<double>& rVariable, const double& rValue,
+                      const ProcessInfo& rCurrentProcessInfo) override;
+
+        void CalculateMohrCoulomb(ConstitutiveLaw::Parameters& parameters);
```
I feel lacking some description of the methods...

avdg81 replied:
As a rule of thumb, we try to name our functions (and their parameters) in such a way that no further description is required. Of course, occasionally we need to clarify a parameter, which we do in that case.

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_constitutive_law.hpp:
```diff
@@ -0,0 +1,107 @@
+        std::unique_ptr<ConstitutiveLawDimension> mpConstitutiveDimension;
+        Vector mStressVector;
+        Vector mStressVectorFinalized;
+        Vector mStrainVectorFinalized;
```
why duplicating the member variables? You can update the member variables just once at the `FinalizeMaterialResponse`

avdg81 replied:
Thanks for your suggestion. We may reconsider it in the future. For now, we have decided to keep the code consistent with existing code in the GeoMechanicsApplication.

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.cpp:
```diff
@@ -176,4 +176,46 @@ std::vector<Vector> StressStrainUtilities::CalculateStrains(const std::vector<Ma
     return result;
 }
 
+void StressStrainUtilities::CalculatePrincipalStresses(const Vector& rCauchyStressVector,
```
also implemented already and tested

avdg81 replied:
We have checked the corresponding function in the ConstitutiveLawApplication, and found that it only returns the principal stress values, but not the principal directions (and we need those). Therefore, we have written our own implementation that returns both.

AlejandroCornejo reviewed (commented):
Just some recommendations taking into account my experience in the `ConstitutiveLawsApp`

mnabideltares committed:
Addressed some of the code smells

mnabideltares committed:
Fix in unit test

mnabideltares committed:
code smells

mnabideltares committed:
Removed a test case from the unit test

mnabideltares committed:
Modifications based on review

mnabideltares committed:
Changes based on review

mnabideltares committed:
modifications

mnabideltares committed:
Merge remote-tracking branch 'origin/master' into geo/13071-implement-mohr-coulomb-continuum

mnabideltares committed:
Modifications and clang format

mnabideltares committed:
fixes

mnabideltares committed:
fixes

mnabideltares committed:
Fixes

mnabideltares committed:
refactoring

mnabideltares committed:
Refactoring

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/yield_surface.h:
```diff
@@ -0,0 +1,29 @@
+
+#pragma once
+
+#include "includes/serializer.h"
```
This `#include` is not needed here. Later, when we add the `load` and `save` member functions, we should use a forward declaration. That is, it should be done in the same way as for class `StressStatePolicy`.

However, we do need `#include "includes/ublas_interface.h"`, since the class uses `Vector` as the return type of one of the member functions.

```suggestion
#include "includes/ublas_interface.h"
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/yield_surface.h:
```diff
@@ -0,0 +1,29 @@
+    virtual ~YieldSurface() = default;
+
+    virtual double YieldFunctionValue(const Vector& rPrincipalStress) const     = 0;
+    virtual Vector DerivateOfFlowFunction(const Vector& rPrincipalStress) const = 0;
```
As suggested by Clang-Tidy, mark these two member function `[[nodiscard]]`:
```suggestion
    [[nodiscard]] virtual double YieldFunctionValue(const Vector& rPrincipalStress) const     = 0;
    [[nodiscard]] virtual Vector DerivateOfFlowFunction(const Vector& rPrincipalStress) const = 0;
```

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/math_utilities.h:
```diff
@@ -36,6 +36,15 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) GeoMechanicsMathUtilities
+        for (std::size_t i = 0; i < rVector.size(); ++i) {
+            result(i, i) = rVector(i);
+        }
+        return result;
```
Since this is a non-template member function, please put the implementation in the corresponding `.cpp` file. In that way, when you make changes to the implementation, you need to recompile less code.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/tension_cutoff.hpp:
Please rename this file to `tension_cutoff.h`. We use the `.hpp` file extension exclusively for class templates and function templates that don't have a corresponding `.cpp` file. In all other cases, we use the `.h` file extension for header files.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/tension_cutoff.hpp:
```diff
@@ -0,0 +1,40 @@
+//  License:         geo_mechanics_application/license.txt
+//
+//
+//  Main authors:    Mohame Nabi,
```
There is a typo in your name:
```suggestion
//  Main authors:    Mohamed Nabi,
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/tension_cutoff.hpp:
```diff
@@ -0,0 +1,40 @@
+//
+//  Main authors:    Mohame Nabi,
+//                   Wijtze Pieter Kikstra
+// //
```
There is a pair of redundant comment markers:
```suggestion
//
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/tension_cutoff.hpp:
```diff
@@ -0,0 +1,40 @@
+
+    TensionCutoff() = default;
+
+    explicit TensionCutoff(double tensileStrength);
```
The parameter name doesn't comply with the Kratos Style Guide:
```suggestion
    explicit TensionCutoff(double TensileStrength);
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/tension_cutoff.hpp:
```diff
@@ -0,0 +1,40 @@
+    Vector DerivateOfFlowFunction(const Vector& rPrincipalStress) const override;
+
+private:
+    // Member Variables
```
This is a redundant comment. Please remove it.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/tension_cutoff.hpp:
```diff
@@ -0,0 +1,40 @@
+    // Member Variables
+    double mTensileStrength = 0.0;
+
+}; // Class TensionCutoffFunction
```
```suggestion
}; // Class TensionCutoff
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/tension_cutoff.hpp:
```diff
@@ -0,0 +1,40 @@
+    explicit TensionCutoff(double tensileStrength);
+
+    double YieldFunctionValue(const Vector& rPrincipalStress) const override;
+    Vector DerivateOfFlowFunction(const Vector& rPrincipalStress) const override;
```
```suggestion
    [[nodiscard]] double YieldFunctionValue(const Vector& rPrincipalStress) const override;
    [[nodiscard]] Vector DerivateOfFlowFunction(const Vector& rPrincipalStress) const override;
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_yield_surface.hpp:
Please, change the file extension to `.h` rather than `.hpp`.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_yield_surface.hpp:
```diff
@@ -0,0 +1,42 @@
+//  License:         geo_mechanics_application/license.txt
+//
+//
+//  Main authors:    Mohame Nabi,
```
```suggestion
//  Main authors:    Mohamed Nabi,
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_yield_surface.hpp:
```diff
@@ -0,0 +1,42 @@
+
+    CoulombYieldSurface() = default;
+
+    CoulombYieldSurface(double frictionAngle, double cohesion, double dilatationAngle);
```
The function parameter names don't follow the Kratos Style Guide:
```suggestion
    CoulombYieldSurface(double FrictionAngle, double Cohesion, double DilatationAngle);
```

Also, for clarity, you may want to add the expected angular units to the function parameter names, e.g. `FrictionAngleInRad`. The same applies to the corresponding data members (`mFrictionAngle` and `mDilatationAngle`).

avdg81 replied:
I still miss the angular unit in the variable names. Could you please add that? Thanks.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_yield_surface.hpp:
```diff
@@ -0,0 +1,42 @@
+    CoulombYieldSurface(double frictionAngle, double cohesion, double dilatationAngle);
+
+    double YieldFunctionValue(const Vector& rPrincipalStress) const override;
+    Vector DerivateOfFlowFunction(const Vector& rPrincipalStress) const override;
```
```suggestion
    [[nodiscard]] double YieldFunctionValue(const Vector& rPrincipalStress) const override;
    [[nodiscard]] Vector DerivateOfFlowFunction(const Vector& rPrincipalStress) const override;
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_yield_surface.hpp:
```diff
@@ -0,0 +1,42 @@
+    Vector DerivateOfFlowFunction(const Vector& rPrincipalStress) const override;
+
+private:
+    // Member Variables
```
Please remove this redundant comment.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_yield_surface.hpp:
```diff
@@ -0,0 +1,42 @@
+    double mCohesion        = 0.0;
+    double mDilatationAngle = 0.0;
+
+}; // Class CoulombYieldFunction
```
```suggestion
}; // Class CoulombYieldSurface
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_yield_surface.cpp:
```diff
@@ -0,0 +1,39 @@
+
+#include "custom_constitutive/coulomb_yield_surface.hpp"
+
+#include <boost/numeric/ublas/assignment.hpp>
```
I would have expected an include for using `std::sin` and `std::cos`:
```suggestion
#include <boost/numeric/ublas/assignment.hpp>
#include <cmath>
```

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_yield_surface.cpp:
```diff
@@ -0,0 +1,53 @@
+
+KRATOS_TEST_CASE_IN_SUITE(TestCoulombYieldSurface, KratosGeoMechanicsFastSuiteWithoutKernel)
+{
+    double friction_angle = 45.0 * Globals::Pi / 180.0;
```
The following function expresses your intent more directly:
```suggestion
    const auto friction_angle = MathUtils<>::DegreesToRadians(45.0);
```

avdg81 replied:
Could you please use `const auto` here rather than `double`?

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_yield_surface.cpp:
```diff
@@ -0,0 +1,53 @@
+KRATOS_TEST_CASE_IN_SUITE(TestCoulombYieldSurface, KratosGeoMechanicsFastSuiteWithoutKernel)
+{
+    double friction_angle = 45.0 * Globals::Pi / 180.0;
+    double cohesion       = 2.0;
```
Make data immutable whenever you can:
```suggestion
    constexpr auto cohesion       = 2.0;
```

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_yield_surface.cpp:
```diff
@@ -0,0 +1,53 @@
+    CoulombYieldSurface coulombYieldSurface(friction_angle, cohesion, 0.0);
+    double              yield_value = coulombYieldSurface.YieldFunctionValue(principal_stress);
+    double              expected_solution = 1.0;
+    KRATOS_EXPECT_NEAR(yield_value, expected_solution, tolerance);
```
Instead of defining a local tolerance, I'd suggest to use the one defined in `tests/cpp_tests/test_utilities.h`:
```suggestion
    KRATOS_EXPECT_NEAR(yield_value, expected_solution, Defaults::absolute_tolerance);
```

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_yield_surface.cpp:
```diff
@@ -0,0 +1,53 @@
+namespace Kratos::Testing
+{
+
+KRATOS_TEST_CASE_IN_SUITE(TestCoulombYieldSurface, KratosGeoMechanicsFastSuiteWithoutKernel)
```
This unit test checks just one principal stress state, i.e. one that is inadmissible since the yield function value is positive. Wouldn't it be nice to also test a principal stress state that is on the yield contour (i.e. $`f = 0`$), and one that is in the elastic region (i.e. $`f < 0`$)?

In that case, I would suggest to slightly reorder the test code:
```cpp
KRATOS_TEST_CASE_IN_SUITE(TestCoulombYieldSurface, KratosGeoMechanicsFastSuiteWithoutKernel)
{
    const auto friction_angle   = MathUtils<>::DegreesToRadians(45.0);
    const auto cohesion         = 2.0;
    const auto dilatation_angle = MathUtils<>::DegreesToRadians(0.0)
    CoulombYieldSurface coulombYieldSurface(friction_angle, cohesion, dilatation_angle);

    Vector principal_stress(3);
    principal_stress <<= 3.0, 2.0, 1.0;
    KRATOS_EXPECT_NEAR(coulombYieldSurface.YieldFunctionValue(principal_stress), 1.0, Defaults::absolute_tolerance);

    // Test here a principal stress state that yields f = 0.0
    principal_stress <<= ...;
    KRATOS_EXPECT_NEAR(...)

    // Test here a principal stress state that yields f < 0.0
    principal_stress <<= ...;
    KRATOS_EXPECT_NEAR(...)
}
```

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_yield_surface.cpp:
```diff
@@ -0,0 +1,53 @@
+    double        tension_value     = tensionCutoff.YieldFunctionValue(principal_stress);
+    double        expected_solution = 1.0;
+    KRATOS_EXPECT_NEAR(tension_value, expected_solution, tolerance);
+}
```
And similar changes here:
```suggestion
KRATOS_TEST_CASE_IN_SUITE(TestTensionCutoff, KratosGeoMechanicsFastSuiteWithoutKernel)
{
    constexpr auto tensile_strength = 2.0;
    TensionCutoff tension_cutoff(tensile_strength);

    Vector principal_stress(3);
    principal_stress <<= 3.0, 2.0, 1.0;
    KRATOS_EXPECT_NEAR(tension_cutoff.YieldFunctionValue(principal_stress), 1.0, Defaults::absolute_tolerance);
    
    // Test here a principal stress state that yields f = 0.0
    principal_stress <<= ...;
    KRATOS_EXPECT_NEAR(...)

    // Test here a principal stress state that yields f < 0.0
    principal_stress <<= ...;
    KRATOS_EXPECT_NEAR(...)
}
```

avdg81 replied:
Not all of my suggested changes have been processed. Could you please have another look? Thanks.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_utilities/test_stress_strain_utitlities.cpp:
```diff
@@ -190,4 +192,62 @@ KRATOS_TEST_CASE_IN_SUITE(CheckCalculateStrains, KratosGeoMechanicsFastSuiteWith
+
+    for (unsigned int i = 0; i < principal_stresses.size(); ++i) {
+        KRATOS_EXPECT_NEAR(principal_stresses(i), expected_solution(i), tolerance);
+    }
```
I think we can simplify this a bit:
```suggestion
    Vector expected_solution(3);
    expected_solution <<= 135.736961146391, 22.5224297324582, -8.25939087884923;
    KRATOS_EXPECT_VECTOR_NEAR(principal_stresses, expected_solution, Defaults::absolute_tolerance)
```

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_utilities/test_stress_strain_utitlities.cpp:
```diff
@@ -190,4 +192,62 @@ KRATOS_TEST_CASE_IN_SUITE(CheckCalculateStrains, KratosGeoMechanicsFastSuiteWith
+    cauchy_stresses <<= 80.0, 50.0, 20.0, 40.0, 35.0, 45.0;
+
+    Vector principal_stresses;
+    Matrix eigenvectors_matrix;
```
Is there any particular reason why you don't validate the calculated eigenvectors?

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_utilities/test_stress_strain_utitlities.cpp:
```diff
@@ -190,4 +192,62 @@ KRATOS_TEST_CASE_IN_SUITE(CheckCalculateStrains, KratosGeoMechanicsFastSuiteWith
+
+KRATOS_TEST_CASE_IN_SUITE(CheckCalculateRotationMatrix2D, KratosGeoMechanicsFastSuiteWithoutKernel)
+{
+    // Set
```
??

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_utilities/test_stress_strain_utitlities.cpp:
```diff
@@ -190,4 +192,62 @@ KRATOS_TEST_CASE_IN_SUITE(CheckCalculateStrains, KratosGeoMechanicsFastSuiteWith
+    }
+}
+
+KRATOS_TEST_CASE_IN_SUITE(CheckCalculateRotationMatrix2D, KratosGeoMechanicsFastSuiteWithoutKernel)
```
By just looking at your unit test, I wouldn't be convinced that the implementation of `StressStrainUtilities::CalculateRotationMatrix` is correct. For instance, suppose that it always returns the identity matrix $`I`$, no matter the which set of eigenvectors you give it. Then this unit test would still pass, since $`I \cdot I^T = I`$.

Perhaps we can improve the test as follows:
- Start from a given set of eigenvectors, rather than calculating them from a given Cauchy stress vector. It makes the test dependent on the correctness of `StressStrainUtilities::CalculatePrincipalStresses`, even though this dependency is not needed.
- I would check the returned rotation matrix itself, rather than the matrix product $`R \cdot R^T`$ for the reason given above.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_utilities/test_stress_strain_utitlities.cpp:
```diff
@@ -190,4 +192,62 @@ KRATOS_TEST_CASE_IN_SUITE(CheckCalculateStrains, KratosGeoMechanicsFastSuiteWith
+    // Set
+    auto   plane_strain  = std::make_unique<PlaneStrain>();
+    Vector stress_vector = ZeroVector(plane_strain->GetStrainSize());
+    stress_vector <<= -6.0, -7.0, -14.0, 1.0;
```
I think using class `PlaneStrain` is overkill here. When you fill `stress_vector`, you know that pass it four values.
```suggestion
    Vector stress_vector(4);
    stress_vector <<= -6.0, -7.0, -14.0, 1.0;
```

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_utilities/test_stress_strain_utitlities.cpp:
```diff
@@ -190,4 +192,62 @@ KRATOS_TEST_CASE_IN_SUITE(CheckCalculateStrains, KratosGeoMechanicsFastSuiteWith
+    KRATOS_EXPECT_MATRIX_NEAR(unit_matrix, expected_unit_matrix, 1.0e-15);
+}
+
+KRATOS_TEST_CASE_IN_SUITE(CheckCalculateRotationMatrix3D, KratosGeoMechanicsFastSuiteWithoutKernel)
```
Could you please propagate any changes that you've made to the previous unit test also the this one?

avdg81 commented on applications/GeoMechanicsApplication/geo_mechanics_application_variables.h:
I don't think we need these changes yet. Let's introduce them when we need them, and add tests to check whether it works.

avdg81 commented on applications/GeoMechanicsApplication/geo_mechanics_application_variables.cpp:
I don't think we need these changes yet. Let's introduce them when we need them, and add tests to check whether it works.

avdg81 commented on applications/GeoMechanicsApplication/geo_mechanics_application.h:
```diff
@@ -125,6 +125,7 @@
 #include "custom_constitutive/linear_elastic_2D_interface_law.h"
 #include "custom_constitutive/linear_elastic_3D_interface_law.h"
 #include "custom_constitutive/linear_elastic_plane_stress_2D_law.h"
+#include "custom_constitutive/mohr_coulomb_with_tension_cutoff.hpp"
```
Not needed yet, so please revert this change.

avdg81 commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
I don't think we need these changes yet. Let's introduce them when we need them, and add tests to check whether it works.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_utilities/test_stress_strain_utitlities.cpp:
I miss a unit test for `RotateStressMatrix`. Can you please add it?

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.h:
```diff
@@ -40,6 +40,11 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) StressStrainUtilities
+    static void                CalculatePrincipalStresses(const Vector& rCauchyStressVector,
+                                                          Vector&       rPrincipalStressVector,
+                                                          Matrix&       rEigenVectorsMatrix);
+    static Matrix              CalculateRotationMatrix(const Matrix& eigenVectorsMatrix);
```
```suggestion
    static Matrix              CalculateRotationMatrix(const Matrix& rEigenVectorsMatrix);
```

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.h:
```diff
@@ -40,6 +40,11 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) StressStrainUtilities
+                                                          Vector&       rPrincipalStressVector,
+                                                          Matrix&       rEigenVectorsMatrix);
+    static Matrix              CalculateRotationMatrix(const Matrix& eigenVectorsMatrix);
+    static Vector RotateStressMatrix(const Matrix& rStressMatrix, const Matrix& rRotationMatrix);
```
Do you mean stress tensor? If yes, then perhaps this is a more general function that rotates second order tensors?

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.cpp:
```diff
@@ -176,4 +177,71 @@ std::vector<Vector> StressStrainUtilities::CalculateStrains(const std::vector<Ma
+    KRATOS_INFO("CalculatePrincipalStresses")
+        << "rPrincipalStressVector (1): " << rPrincipalStressVector << std::endl;
+    KRATOS_INFO("CalculatePrincipalStresses")
+        << "rEigenVectorsMatrix (1): " << rEigenVectorsMatrix << std::endl;
```
These are debugging statements that should not be in production code.

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.cpp:
```diff
@@ -176,4 +177,71 @@ std::vector<Vector> StressStrainUtilities::CalculateStrains(const std::vector<Ma
+        << "rEigenVectorsMatrix (1): " << rEigenVectorsMatrix << std::endl;
+    std::vector<std::size_t> indices(rPrincipalStressVector.size());
+    std::iota(indices.begin(), indices.end(), 0);
+    KRATOS_INFO("CalculatePrincipalStresses") << "indices (1): " << indices << std::endl;
```
Another debugging statement.

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.cpp:
```diff
@@ -176,4 +177,71 @@ std::vector<Vector> StressStrainUtilities::CalculateStrains(const std::vector<Ma
+    });
+    KRATOS_INFO("CalculatePrincipalStresses")
+        << "rPrincipalStressVector (2): " << rPrincipalStressVector << std::endl;
+    KRATOS_INFO("CalculatePrincipalStresses") << "indices (2): " << indices << std::endl;
```
Even more debugging statements.

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.cpp:
```diff
@@ -176,4 +177,71 @@ std::vector<Vector> StressStrainUtilities::CalculateStrains(const std::vector<Ma
+
+    std::copy(tmp_vector.begin(), tmp_vector.end(), rPrincipalStressVector.begin());
+    KRATOS_INFO("CalculatePrincipalStresses")
+        << "rPrincipalStressVector (3): " << rPrincipalStressVector << std::endl;
```
Debugging statement.

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.cpp:
```diff
@@ -176,4 +177,71 @@ std::vector<Vector> StressStrainUtilities::CalculateStrains(const std::vector<Ma
+
+    noalias(rEigenVectorsMatrix) = tmp_matrix;
+    KRATOS_INFO("CalculatePrincipalStresses")
+        << "rEigenVectorsMatrix (3): " << rEigenVectorsMatrix << std::endl;
```
Debugging statement.

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.cpp:
```diff
@@ -176,4 +177,71 @@ std::vector<Vector> StressStrainUtilities::CalculateStrains(const std::vector<Ma
+                                                       Vector&       rPrincipalStressVector,
+                                                       Matrix&       rEigenVectorsMatrix)
+{
+    auto   stress_tensor = MathUtils<double>::StressVectorToTensor(rCauchyStressVector);
```
```suggestion
    const auto   stress_tensor = MathUtils<>::StressVectorToTensor(rCauchyStressVector);
```

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.cpp:
```diff
@@ -176,4 +177,71 @@ std::vector<Vector> StressStrainUtilities::CalculateStrains(const std::vector<Ma
+{
+    auto   stress_tensor = MathUtils<double>::StressVectorToTensor(rCauchyStressVector);
+    Matrix PrincipalStressMatrix;
+    Matrix EigenVectorsMatrix;
```
These names of local variables don't comply with the Kratos Style Guide:
```suggestion
    Matrix principal_stress_tensor;
    Matrix eigen_vectors;
```

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.cpp:
```diff
@@ -176,4 +177,71 @@ std::vector<Vector> StressStrainUtilities::CalculateStrains(const std::vector<Ma
+    KRATOS_INFO("CalculatePrincipalStresses")
+        << "rEigenVectorsMatrix (1): " << rEigenVectorsMatrix << std::endl;
+    std::vector<std::size_t> indices(rPrincipalStressVector.size());
+    std::iota(indices.begin(), indices.end(), 0);
```
```suggestion
    std::iota(indices.begin(), indices.end(), std::size_t{0});
```

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.cpp:
```diff
@@ -176,4 +177,71 @@ std::vector<Vector> StressStrainUtilities::CalculateStrains(const std::vector<Ma
+    KRATOS_INFO("CalculatePrincipalStresses") << "indices (1): " << indices << std::endl;
+    std::sort(indices.begin(), indices.end(), [&rPrincipalStressVector](const std::size_t i, const std::size_t j) {
+        return rPrincipalStressVector[j] < rPrincipalStressVector[i];
+    });
```
```suggestion
    std::sort(indices.begin(), indices.end(), [&rPrincipalStressVector](auto i, auto j) {
        return rPrincipalStressVector[j] < rPrincipalStressVector[i];
    });
```

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.cpp:
```diff
@@ -176,4 +177,71 @@ std::vector<Vector> StressStrainUtilities::CalculateStrains(const std::vector<Ma
+        tmp_vector.push_back(rPrincipalStressVector[index]);
+    }
+
+    std::copy(tmp_vector.begin(), tmp_vector.end(), rPrincipalStressVector.begin());
```
Here is where we apply the permutation to the principal stress vector. Instead of writing this inline, we may consider to move this into a generic function. I've found two online articles that describe the problem and propose solutions to it: https://devblogs.microsoft.com/oldnewthing/20170102-00/?p=95095 and https://blog.merovius.de/posts/2014-08-12-applying-permutation-in-constant/. It seems that neither the STL nor Boost have functions for this.

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.cpp:
```diff
@@ -176,4 +177,71 @@ std::vector<Vector> StressStrainUtilities::CalculateStrains(const std::vector<Ma
+        }
+    }
+
+    noalias(rEigenVectorsMatrix) = tmp_matrix;
```
Also here, we are applying a permutation to the columns of a matrix. It would be nice to encapsulate that in a utility function.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.hpp:
```diff
@@ -0,0 +1,91 @@
+//
+//  Main authors:    Mohamed Nabi,
+//                   Wijtze Pieter Kikstra
+// //
```
```suggestion
//
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.hpp:
```diff
@@ -0,0 +1,91 @@
+// Project includes
+#include "custom_constitutive/coulomb_yield_surface.hpp"
+#include "custom_constitutive/tension_cutoff.hpp"
+#include "geo_mechanics_application_variables.h"
```
Can you explain to me why we need this include here, because I don't see it.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.hpp:
```diff
@@ -0,0 +1,91 @@
+#include "custom_constitutive/tension_cutoff.hpp"
+#include "geo_mechanics_application_variables.h"
+#include "includes/constitutive_law.h"
+#include "includes/serializer.h"
```
We can move this `#include` to the implementation file (`.cpp`), since a forward declaration will do here. See also my next suggestion.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.hpp:
```diff
@@ -0,0 +1,91 @@
+
+namespace Kratos
+{
+class ConstitutiveLawDimension;
```
Add a forward declaration here:
```suggestion
class ConstitutiveLawDimension;
class Serializer;
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.hpp:
```diff
@@ -0,0 +1,91 @@
+    int  Check(const Properties&   rMaterialProperties,
+               const GeometryType& rElementGeometry,
+               const ProcessInfo&  rCurrentProcessInfo) const override;
+    void CalculateMaterialResponseCauchy(ConstitutiveLaw::Parameters& parameters) override;
```
Kratos Style Guide:
```suggestion
    void CalculateMaterialResponseCauchy(ConstitutiveLaw::Parameters& rParameters) override;
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.hpp:
```diff
@@ -0,0 +1,91 @@
+    void FinalizeMaterialResponseCauchy(ConstitutiveLaw::Parameters& rValues) override;
+    Vector RotatePrincipalStresses(const Vector& rPrincipalStressVector, const Matrix& rRotationMatrix) const;
+
+    double FindRegion(ConstitutiveLaw::Parameters& rParameters, const Vector& rPrincipalStressVector);
```
These sound like implementation details to me, and therefore shouldn't be in the `public` interface of this class. In fact, `FindRegion` has not been implemented at all. It's probably a leftover from a previous attempt, so please remove its declaration here.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.hpp:
```diff
@@ -0,0 +1,91 @@
+    friend class Serializer;
+    void save(Serializer& rSerializer) const override;
+    void load(Serializer& rSerializer) override;
+}; // Class MohrCoulombConstitutiveLaw
```
```suggestion
}; // Class MohrCoulombWithTensionCutOff
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.hpp:
Please use file extension `.h` rather than `.hpp`.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+// Application includes
+#include "custom_constitutive/mohr_coulomb_with_tension_cutoff.hpp"
+#include "custom_constitutive/constitutive_law_dimension.h"
+#include "custom_constitutive/tension_cutoff.hpp"
```
This file is already included by `custom_constitutive/mohr_coulomb_with_tension_cutoff.hpp`, so there is no need to include it once more here.
```suggestion
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+
+    CheckProperty(rMaterialProperties, GEO_COHESION);
+    CheckProperty(rMaterialProperties, GEO_FRICTION_ANGLE);
+    CheckProperty(rMaterialProperties, GEO_DILATION_ANGLE);
```
Isn't the official term "dilatancy angle"?

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+    CheckProperty(rMaterialProperties, GEO_COHESION);
+    CheckProperty(rMaterialProperties, GEO_FRICTION_ANGLE);
+    CheckProperty(rMaterialProperties, GEO_DILATION_ANGLE);
+    CheckProperty(rMaterialProperties, GEO_TENSION_CUTOFF);
```
Don't you mean tensile strength?

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+void MohrCoulombWithTensionCutOff::CalculateMaterialResponseCauchy(ConstitutiveLaw::Parameters& rParameters)
+{
+    const Properties& r_prop         = rParameters.GetMaterialProperties();
+    const auto        friction_angle = r_prop[GEO_FRICTION_ANGLE] * Globals::Pi / 180.0;
```
```suggestion
    const auto        friction_angle_in_rad = MathUtils<>::DegreesToRadians(r_prop[GEO_FRICTION_ANGLE]);
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+    const Properties& r_prop         = rParameters.GetMaterialProperties();
+    const auto        friction_angle = r_prop[GEO_FRICTION_ANGLE] * Globals::Pi / 180.0;
+    const auto        cohesion       = r_prop[GEO_COHESION];
+    const auto        dilation_angle = r_prop[GEO_DILATION_ANGLE] * Globals::Pi / 180.0;
```
```suggestion
    const auto        dilatancy_angle = MathUtils::DegreesToRadians(r_prop[GEO_DILATANCY_ANGLE]);
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+    const auto        friction_angle = r_prop[GEO_FRICTION_ANGLE] * Globals::Pi / 180.0;
+    const auto        cohesion       = r_prop[GEO_COHESION];
+    const auto        dilation_angle = r_prop[GEO_DILATION_ANGLE] * Globals::Pi / 180.0;
+    const auto        tension_cutoff = r_prop[GEO_TENSION_CUTOFF];
```
```suggestion
    const auto        tensile_strength = r_prop[GEO_TENSILE_STRENGTH];
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+    double cutoff  = mTensionCutOff.YieldFunctionValue(principal_trial_stress_vector);
+
+    // Elastic region
+    if (coulomb <= 0.0 && cutoff <= 0.0) {
```
Shouldn't be just `< 0.0`? I thought that `== 0.0` meant "on the yield contour", where plastic deformation starts or continues.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+    const auto        tension_cutoff = r_prop[GEO_TENSION_CUTOFF];
+
+    mCoulombYieldSurface = CoulombYieldSurface(friction_angle, cohesion, dilation_angle);
+    mTensionCutOff       = TensionCutoff(tension_cutoff);
```
Perhaps we can move this to `InitializeMaterial`, just so we do this only once? If we need to keep it here, then I don't see why they are data members, since they are reconstructed every time.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+    Matrix eigenvectors_matrix;
+    StressStrainUtilities::CalculatePrincipalStresses(
+        trail_stress_vector, principal_trial_stress_vector, eigenvectors_matrix);
+    Matrix rotation_matrix = StressStrainUtilities::CalculateRotationMatrix(eigenvectors_matrix);
```
We use `rotation_matrix` only much later, so I would postpone its calculation until we really need it.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+    Matrix rotation_matrix = StressStrainUtilities::CalculateRotationMatrix(eigenvectors_matrix);
+
+    double coulomb = mCoulombYieldSurface.YieldFunctionValue(principal_trial_stress_vector);
+    double cutoff  = mTensionCutOff.YieldFunctionValue(principal_trial_stress_vector);
```
```suggestion
    const auto coulomb = mCoulombYieldSurface.YieldFunctionValue(principal_trial_stress_vector);
    const auto cutoff  = mTensionCutOff.YieldFunctionValue(principal_trial_stress_vector);
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+    }
+
+    // Tension cut-off return
+    double trail_sigma = 0.5 * (principal_trial_stress_vector(0) + principal_trial_stress_vector(2));
```
It appears we don't need `trail_sigma` yet. Can we perhaps move its calculation close to its first usage? And make it `const` please.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+    double trail_sigma = 0.5 * (principal_trial_stress_vector(0) + principal_trial_stress_vector(2));
+    double trial_tau = 0.5 * (principal_trial_stress_vector(0) - principal_trial_stress_vector(2));
+    double apex      = this->CalculateApex(friction_angle, cohesion);
+    Vector corner_point = this->CalculateCornerPoint(friction_angle, cohesion, tension_cutoff);
```
```suggestion
    const auto trial_tau = 0.5 * (principal_trial_stress_vector(0) - principal_trial_stress_vector(2));
    const auto apex      = this->CalculateApex(friction_angle, cohesion);
    const auto corner_point = this->CalculateCornerPoint(friction_angle, cohesion, tension_cutoff);
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+}
+
+Vector MohrCoulombWithTensionCutOff::ReturnStressAtAxialZone(const Vector& rPrincipalTrialStressVector,
+                                                             double TensionCutoff) const
```
```suggestion
                                                             double TensileStrength) const
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+                                                                      double Cohesion) const
+{
+    Vector result = rPrincipalTrialStressVector;
+    Vector flowDerivative = rCoulombYieldSurface.DerivateOfFlowFunction(rPrincipalTrialStressVector);
```
```suggestion
    const auto flow_derivative = rCoulombYieldSurface.DerivateOfFlowFunction(rPrincipalTrialStressVector);
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+    Vector flowDerivative = rCoulombYieldSurface.DerivateOfFlowFunction(rPrincipalTrialStressVector);
+    double cof1   = (1.0 + std::sin(FrictionAngle)) / (1.0 - std::sin(FrictionAngle));
+    double cof2   = 2.0 * Cohesion * std::cos(FrictionAngle) / (1.0 - std::sin(FrictionAngle));
+    double det    = cof1 * flowDerivative(0) - flowDerivative(2);
```
```suggestion
    const auto cof1   = (1.0 + std::sin(FrictionAngle)) / (1.0 - std::sin(FrictionAngle));
    const auto cof2   = 2.0 * Cohesion * std::cos(FrictionAngle) / (1.0 - std::sin(FrictionAngle));
    const auto det    = cof1 * flowDerivative(0) - flowDerivative(2);
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+    double cof2   = 2.0 * Cohesion * std::cos(FrictionAngle) / (1.0 - std::sin(FrictionAngle));
+    double det    = cof1 * flowDerivative(0) - flowDerivative(2);
+    double lambda = rPrincipalTrialStressVector(2) + cof2 - rPrincipalTrialStressVector(0) * cof1;
+    lambda /= det;
```
```suggestion
    const auto lambda = (rPrincipalTrialStressVector(2) + cof2 - rPrincipalTrialStressVector(0) * cof1) / det;
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+    return Cohesion / std::tan(FrictionAngle);
+}
+
+Vector MohrCoulombWithTensionCutOff::CalculateCornerPoint(double FrictionAngle, double Cohesion, double TensionCutoff) const
```
```suggestion
Vector MohrCoulombWithTensionCutOff::CalculateCornerPoint(double FrictionAngle, double Cohesion, double TensileStrength) const
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+    Matrix principal_stress_matrix = GeoMechanicsMathUtilities::VectorToDiagonalMatrix(rPrincipalStressVector);
+    Vector result = StressStrainUtilities::RotateStressMatrix(principal_stress_matrix, rRotationMatrix);
+    return result;
+}
```
Perhaps we'd better move this to a utility function?

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+                                                              Vector&       rStressVector,
+                                                              ConstitutiveLaw::Parameters& rValues) const
+{
+    Vector deltaStrainVector = rValues.GetStrainVector() - mStrainVectorFinalized;
```
```suggestion
    const auto delta_strain_vector = rStrainVector - mStrainVectorFinalized;
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+
+void MohrCoulombWithTensionCutOff::CalculateTrialStressVector(const Vector& rStrainVector,
+                                                              Vector&       rStressVector,
+                                                              ConstitutiveLaw::Parameters& rValues) const
```
This parameter is never used. That is very confusing. What about using this parameter and letting the caller provide it (e.g. by calling `rValues.GetStrainVector`)?

Also, I would suggest to use a proper return type here, rather than passing its result back through the parameter list:
```suggestion
Vector MohrCoulombWithTensionCutOff::CalculateTrialStressVector(const Vector& rStrainVector,
                                                                ConstitutiveLaw::Parameters& rValues) const
```

In addition, we could also supply Young's modulus and Poisson's ratio rather than the full `ConstitutiveLaw::Parameters`.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,246 @@
+    const Properties& r_material_properties = rValues.GetMaterialProperties();
+    return mpConstitutiveDimension->CalculateElasticMatrix(r_material_properties[YOUNG_MODULUS],
+                                                           r_material_properties[POISSON_RATIO]);
+}
```
When you supply Young's modulus and Poisson's ratio directly to `CalculateTrialStressVector`, I think we no longer need this member function. We can simply inline its function body.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,115 @@
+
+KRATOS_TEST_CASE_IN_SUITE(MohrCoulombWithTensionCutOff_Clone, KratosGeoMechanicsFastSuiteWithoutKernel)
+{
+    // Set
```
It is more common to use "Arrange":
```suggestion
    // Arrange
```

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,115 @@
+                          KratosGeoMechanicsFastSuiteWithoutKernel)
+{
+    // Set
+    auto plane_strain = std::make_unique<PlaneStrain>();
```
This variable is being used only to determine the sizes of vectors. I would simply hard-code the size (i.e. 4), since the next thing you do after vector construction is filling it with four values.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,115 @@
+
+    // Assert
+    Vector expected_cauchy_stress_vector(plane_strain->GetStrainSize());
+    expected_cauchy_stress_vector <<= -10.0, -10.0, -10.0, 0.0;
```
Does this mean we're in the elastic region? If yes, I would reflect that in the test name.

Also, I would add unit tests to cover the other three regions.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_mohr_coulomb_with_tension_cutoff.cpp:
These unit tests are a great start, but they only cover a part of the public interface. Perhaps we can increase the code coverage (and our confidence in the code) by adding a few more unit tests?

avdg81 reviewed (commented):
Hi Mohamed, thank you very much for picking up the review comments that @WPK4FEM and I have given you last week. You have made quite some progress, which allowed me to look in more detail to the code. As you can see, I have many suggestions, but most of them are minor in nature and require little work to change. Others will require more effort from your side, but as I said before, you're heading into the right direction. Please let me know if things are unclear or when you disagree with my suggestions. We can then discuss them.

mnabideltares committed:
Added unit tests, Added GetValue and SetValue for stress, modified the rotation

mnabideltares committed:
Adresses review comments, part 1

mnabideltares committed:
Adresses review comments, part 2

mnabideltares committed:
Adresses review comments, part 3

mnabideltares committed:
Added documentation

mnabideltares committed:
fixes in Readme

mnabideltares committed:
fix in figure

mnabideltares committed:
fix

mnabideltares committed:
fix

mnabideltares committed:
Restructuring based on review

mnabideltares committed:
Fixed the situation, if after mapping and reordering, it still falls in the inadmissable region.

Added unit test for this too.

mnabideltares committed:
Permutation is added

mnabideltares committed:
Added another unit test

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/yield_surface.h:
```diff
@@ -0,0 +1,29 @@
+    virtual ~YieldSurface() = default;
+
+    [[nodiscard]] virtual double YieldFunctionValue(const Vector& rPrincipalStress) const     = 0;
+    [[nodiscard]] virtual Vector DerivateOfFlowFunction(const Vector& rPrincipalStress) const = 0;
```
```suggestion
    [[nodiscard]] virtual Vector DerivativeOfFlowFunction(const Vector& rPrincipalStress) const = 0;
```

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_yield_surface.cpp:
```diff
@@ -0,0 +1,71 @@
+    principal_stress <<= 3.0, 2.0, 1.0;
+    double yield_value       = coulombYieldSurface.YieldFunctionValue(principal_stress);
+    double expected_solution = 1.0;
+    KRATOS_EXPECT_NEAR(yield_value, expected_solution, Defaults::absolute_tolerance);
```
We don't need to introduce these variables in my opinion. Using their values directly is just as clear and reduces the number of variables needed:
```suggestion
    KRATOS_EXPECT_NEAR(coulombYieldSurface.YieldFunctionValue(principal_stress), 1.0, Defaults::absolute_tolerance);
```

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_yield_surface.cpp:
```diff
@@ -0,0 +1,71 @@
+    double           friction_angle = MathUtils<>::DegreesToRadians(45.0);
+    constexpr double cohesion       = 2.0;
+
+    CoulombYieldSurface coulombYieldSurface(friction_angle, cohesion, 0.0);
```
According to the Kratos Style Guide:
```suggestion
    constexpr auto dilatancy_angle = MathUtils<>::DegreesToRadians(0.0);
    CoulombYieldSurface coulomb_yield_surface(friction_angle, cohesion, dilatancy_angle);
```

Also, I would give a name to the value of the final parameter (e.g. `dilatancy_angle`).

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_yield_surface.cpp:
```diff
@@ -0,0 +1,71 @@
+
+KRATOS_TEST_CASE_IN_SUITE(TestCoulombYieldSurface, KratosGeoMechanicsFastSuiteWithoutKernel)
+{
+    double           friction_angle = MathUtils<>::DegreesToRadians(45.0);
```
```suggestion
    const auto friction_angle = MathUtils<>::DegreesToRadians(45.0);
```

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/math_utilities.h:
```diff
@@ -36,6 +36,9 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) GeoMechanicsMathUtilities
     }
 
+    [[nodiscard]] static Matrix VectorToDiagonalMatrix(const Vector& rVector);
+    [[nodiscard]] static Matrix RotateTensor(const Matrix& rTensor, const Matrix& rRotationMatrix);
```
Perhaps we should make this function name slightly more specific:
```suggestion
    [[nodiscard]] static Matrix RotateSecondOrderTensor(const Matrix& rTensor, const Matrix& rRotationMatrix);
```

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/math_utilities.cpp:
```diff
@@ -15,11 +15,31 @@
+Matrix GeoMechanicsMathUtilities::RotateTensor(const Matrix& rTensor, const Matrix& rRotationMatrix)
+{
+    Matrix temp = prod(rTensor, trans(rRotationMatrix));
+    return prod(rRotationMatrix, temp);
```
We can rewrite this as a one-liner:
```suggestion
    return prod(rRotationMatrix, Matrix{prod(rTensor, trans(rRotationMatrix))});
```

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_utilities/test_math_utilities.cpp:
```diff
@@ -58,4 +60,30 @@ KRATOS_TEST_CASE_IN_SUITE(Normalized_Throws_WhenInputtingZeroVector, KratosGeoMe
+KRATOS_TEST_CASE_IN_SUITE(CheckRotateTensor, KratosGeoMechanicsFastSuiteWithoutKernel)
+{
+    // Arrange
+    Matrix stress_matrix = ZeroMatrix(3, 3);
```
Wouldn't `stress_tensor` be a better name?

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_utilities/test_math_utilities.cpp:
```diff
@@ -58,4 +60,30 @@ KRATOS_TEST_CASE_IN_SUITE(Normalized_Throws_WhenInputtingZeroVector, KratosGeoMe
+                      0.0, 0.0, 20.0;
+    // clang-format on
+
+    double angle           = MathUtils<>::DegreesToRadians(30.0);
```
```suggestion
    const auto angle           = MathUtils<>::DegreesToRadians(30.0);
```

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_utilities/test_math_utilities.cpp:
```diff
@@ -58,4 +60,30 @@ KRATOS_TEST_CASE_IN_SUITE(Normalized_Throws_WhenInputtingZeroVector, KratosGeoMe
+    Matrix rotation_matrix = ZeroMatrix(3, 3);
+    rotation_matrix <<= std::cos(angle), -std::sin(angle), 0.0, std::sin(angle), std::cos(angle),
+        0.0, 0.0, 0.0, 1.0;
+    Matrix result = GeoMechanicsMathUtilities::RotateTensor(stress_matrix, rotation_matrix);
```
```suggestion
    const auto result = GeoMechanicsMathUtilities::RotateTensor(stress_matrix, rotation_matrix);
```

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.h:
```diff
@@ -40,6 +40,14 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) StressStrainUtilities
+    static void                CalculatePrincipalStresses(const Vector& rCauchyStressVector,
+                                                          Vector&       rPrincipalStressVector,
+                                                          Matrix&       rEigenVectorsMatrix);
+    static Vector RotateStressMatrix(const Matrix& rStressMatrix, const Matrix& rRotationMatrix, std::size_t StressVectorSize);
```
This member function is declared, but not defined. It should be removed.

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/generic_utilities.h:
```diff
@@ -0,0 +1,41 @@
+            result[i] = vec[indices[i]];
+        }
+        return result;
+    }
```
We've discussed how this can be improved, and you've received my patch.

mnabideltares committed:
permutation is restructured

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.cpp:
```diff
@@ -176,4 +178,41 @@ std::vector<Vector> StressStrainUtilities::CalculateStrains(const std::vector<Ma
+{
+    Matrix principal_stress_matrix = GeoMechanicsMathUtilities::VectorToDiagonalMatrix(rPrincipalStressVector);
+    Matrix rotated_stress_matrix =
+        GeoMechanicsMathUtilities::RotateTensor(principal_stress_matrix, rRotationMatrix);
```
```suggestion
    const auto principal_stress_matrix = GeoMechanicsMathUtilities::VectorToDiagonalMatrix(rPrincipalStressVector);
    const auto rotated_stress_matrix =
        GeoMechanicsMathUtilities::RotateTensor(principal_stress_matrix, rRotationMatrix);
```

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.cpp:
```diff
@@ -176,4 +178,41 @@ std::vector<Vector> StressStrainUtilities::CalculateStrains(const std::vector<Ma
+    auto   stress_tensor = MathUtils<>::StressVectorToTensor(rCauchyStressVector);
+    Matrix principal_stress_matrix;
+    MathUtils<double>::GaussSeidelEigenSystem(stress_tensor, rEigenVectorsMatrix,
+                                              principal_stress_matrix, 1.0e-16, 20);
```
```suggestion
    Matrix principal_stress_matrix;
    MathUtils<>::GaussSeidelEigenSystem(MathUtils<>::StressVectorToTensor(rCauchyStressVector),
                                        rEigenVectorsMatrix, principal_stress_matrix, 1.0e-16, 20);
```

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.cpp:
```diff
@@ -176,4 +178,41 @@ std::vector<Vector> StressStrainUtilities::CalculateStrains(const std::vector<Ma
+    rPrincipalStressVector = ZeroVector(3);
+    for (int i = 0; i < 3; ++i) {
+        rPrincipalStressVector(i) = principal_stress_matrix(i, i);
+    }
```
Now that you have implemented `VectorToDiagonalMatrix`, you could also implement the inverse operation, i.e. `DiagonalMatrixToVector` in class `GeoMechanicsMathUtilities`.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.h:
```diff
@@ -0,0 +1,92 @@
+               const ProcessInfo&  rCurrentProcessInfo) const override;
+    void CalculateMaterialResponseCauchy(ConstitutiveLaw::Parameters& rParameters) override;
+    void FinalizeMaterialResponseCauchy(ConstitutiveLaw::Parameters& rValues) override;
+    Vector RotatePrincipalStresses(const Vector& rPrincipalStressVector, const Matrix& rRotationMatrix) const;
```
This member is declared, but not defined. Please remove it.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.h:
Clang-tidy suggests to mark the following member functions `[[nodiscard]]`: `GetStrainSize`, `Check`, `CalculateApex`, `CalculateCornerPoint`, `ReturnStressAtAxialZone`, `ReturnStressAtCornerReturnZone`, and `ReturnStressAtRegularFailureZone`.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_utilities/test_stress_strain_utitlities.cpp:
```diff
@@ -11,9 +11,12 @@
 //
 
+#include "custom_constitutive/plane_strain.h"
+#include "custom_constitutive/three_dimensional.h"
```
We no longer need these `#include`s.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.h:
```diff
@@ -0,0 +1,92 @@
+    CoulombYieldSurface                       mCoulombYieldSurface;
+    TensionCutoff                             mTensionCutOff;
+
+    void CheckProperty(const Properties& rMaterialProperties, const Kratos::Variable<double>& rVariable) const;
```
```suggestion
    void CheckProperty(const Properties& rMaterialProperties, const Variable<double>& rVariable) const;
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.h:
```diff
@@ -0,0 +1,92 @@
+    void   CalculateTrialStressVector(const Vector& rStrainVector,
+                                      Vector&       rStressVector,
+                                      double        YoungModulus,
+                                      double        PoissonRatio) const;
```
```suggestion
                                      double        YoungsModulus,
                                      double        PoissonsRatio) const;
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.h:
```diff
@@ -0,0 +1,92 @@
+                                      double        PoissonRatio) const;
+    double CalculateApex(double FrictionAngle, double Cohesion) const;
+    Vector CalculateCornerPoint(double FrictionAngle, double Cohesion, double TensionCutoff) const;
+    Vector ReturnStressAtAxialZone(const Vector& rPrincipalTrialStressVector, double TensionCutoff) const;
```
```suggestion
    Vector CalculateCornerPoint(double FrictionAngle, double Cohesion, double TensileStrength) const;
    Vector ReturnStressAtAxialZone(const Vector& rPrincipalTrialStressVector, double TensileStrength) const;
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,272 @@
+    const auto friction_angle_in_rad = MathUtils<>::DegreesToRadians(r_prop[GEO_FRICTION_ANGLE]);
+    const auto cohesion              = r_prop[GEO_COHESION];
+    const auto dilatancy_angle       = MathUtils<>::DegreesToRadians(r_prop[GEO_DILATANCY_ANGLE]);
+    const auto tension_cutoff        = r_prop[GEO_TENSILE_STRENGTH];
```
```suggestion
    const auto tensile_strength = r_prop[GEO_TENSILE_STRENGTH];
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,272 @@
+    p_result->mStressVector          = mStressVector;
+    p_result->mStressVectorFinalized = mStressVectorFinalized;
+    p_result->mStrainVectorFinalized = mStrainVectorFinalized;
+    return p_result;
```
We also need to copy `mCoulombYieldSurface` and `mTensionCutoff`:
```suggestion
    p_result->mCoulombYieldSurface = mCoulombYieldSurface;
    p_result->mTensionCutoff = mTensionCutoff;
    return p_result;
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,272 @@
+                                        const GeometryType& rElementGeometry,
+                                        const ProcessInfo&  rCurrentProcessInfo) const
+{
+    ConstitutiveLaw::Check(rMaterialProperties, rElementGeometry, rCurrentProcessInfo);
```
Store the value returned by the base class `Check` and return it at the end.
```suggestion
    const auto result = ConstitutiveLaw::Check(rMaterialProperties, rElementGeometry, rCurrentProcessInfo);
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,272 @@
+    CheckProperty(rMaterialProperties, GEO_TENSILE_STRENGTH);
+    CheckProperty(rMaterialProperties, YOUNG_MODULUS);
+    CheckProperty(rMaterialProperties, POISSON_RATIO);
+    return 0;
```
```suggestion
    return result;
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,272 @@
+{
+    const auto friction_angle_in_rad = MathUtils<>::DegreesToRadians(rMaterialProperties[GEO_FRICTION_ANGLE]);
+    const auto cohesion = rMaterialProperties[GEO_COHESION];
+    const auto dilatancy_angle = MathUtils<>::DegreesToRadians(rMaterialProperties[GEO_DILATANCY_ANGLE]);
```
```suggestion
    const auto dilatancy_angle_in_rad = MathUtils<>::DegreesToRadians(rMaterialProperties[GEO_DILATANCY_ANGLE]);
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,272 @@
+    Vector trail_stress_vector;
+    Vector strain_vector = rParameters.GetStrainVector();
+
+    this->CalculateTrialStressVector(strain_vector, trail_stress_vector, young_modulus, poisson_ratio);
```
```suggestion
    this->CalculateTrialStressVector(rParameters.GetStrainVector(), trail_stress_vector, r_prop[YOUNG_MODULUS], r_prop[POISSON_RATIO]);
```

In this way, we also no longer need the (constant) variables `young_modulus` and `poisson_ratio`.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,272 @@
+                                                 const Kratos::Variable<double>& rVariable) const
+{
+    constexpr auto min_value = 0.0;
+    KRATOS_ERROR_IF_NOT(rMaterialProperties.Has(rVariable) || rMaterialProperties[rVariable] < min_value)
```
Since the `min_value` is used only once, we might as well write it in the condition itself:
```suggestion
    KRATOS_ERROR_IF_NOT(rMaterialProperties.Has(rVariable) || rMaterialProperties[rVariable] < 0.0)
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,272 @@
+                                                                      double Cohesion) const
+{
+    Vector result = rPrincipalTrialStressVector;
+    const auto flowDerivative = rCoulombYieldSurface.DerivateOfFlowFunction(rPrincipalTrialStressVector);
```
Kratos Style Guide:
```suggestion
    const auto flow_derivative = rCoulombYieldSurface.DerivateOfFlowFunction(rPrincipalTrialStressVector);
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,272 @@
+    rSerializer.load("StressVector", mStressVector);
+    rSerializer.load("StressVectorFinalized", mStressVectorFinalized);
+    rSerializer.load("StrainVectorFinalized", mStrainVectorFinalized);
+}
```
These two members miss the (de)serialization of `mCoulombYieldSurface` and `mTensionCutOff`.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,272 @@
+    const auto flowDerivative = rCoulombYieldSurface.DerivateOfFlowFunction(rPrincipalTrialStressVector);
+    const auto cof1 = (1.0 + std::sin(FrictionAngle)) / (1.0 - std::sin(FrictionAngle));
+    const auto cof2 = 2.0 * Cohesion * std::cos(FrictionAngle) / (1.0 - std::sin(FrictionAngle));
+    const auto det  = cof1 * flowDerivative(0) - flowDerivative(2);
```
Perhaps `numerator` is a better name than `det`?

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,272 @@
+    const auto lambda = (rPrincipalTrialStressVector(2) + cof2 - rPrincipalTrialStressVector(0) * cof1) / det;
+    result(0) = lambda * flowDerivative(0) + rPrincipalTrialStressVector(0);
+    result(2) = result(0) * cof1 - cof2;
+    return result;
```
We can simplify this as follows:
```suggestion
    return rPrincipalTrialStressVector + lambda * flowDerivative;
```

You can then also remove the variable `result`.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,272 @@
+    return mpConstitutiveDimension->GetDimension();
+}
+
+int MohrCoulombWithTensionCutOff::Check(const Properties&   rMaterialProperties,
```
We would suggest to add a few more checks:
- The admissible range for $`\psi`$ is [0, $`\phi`$].
- The admissible range for the tensile strength is [0, $`\frac{c}{\tan \phi}`$].

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/README.md:
```diff
@@ -47,3 +47,57 @@ Where:
+<img src="documentation_data/mohr-coulomb-with-tension-cutoff-zones.svg" alt="Mohr-Coulomb with tension cutoff" title="Mohr-Coulomb with tension cutoff" width="800">
+
+
+### Implementation in Numerical Models
```
Perhaps simply "Implementation" is enough?

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/README.md:
```diff
@@ -47,3 +47,57 @@ Where:
+
+1. Calculate the trial stress by: 
+
+$$\boldsymbol{\sigma}^{trial} = \boldsymbol{\sigma}^0 + E \Delta \boldsymbol{\epsilon}$$
```
The common notation for the constitutive tensor is $`\boldsymbol{\mathrm{C}}`$ (or $`\boldsymbol{\mathrm{D}}`$):
```suggestion
$`\boldsymbol{\sigma}^{trial} = \boldsymbol{\sigma}^0 + \boldsymbol{\mathrm{C}} \Delta \boldsymbol{\epsilon}`$
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/README.md:
For your mathematical expression, I would suggest to use pairs of `` $` `` and `` `$ ``. If you want to have a complete block, you can also use ```` ```math ```` and ```` ``` ````.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/README.md:
```diff
@@ -47,3 +47,57 @@ Where:
+
+- $\sigma_1$ = maximum principal stress component
+- $\sigma_3$ = minimum principal stress component
+- $C$ = cohesion
```
For cohesion, we usually use the lowercase $`c`$.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/README.md:
```diff
@@ -47,3 +47,57 @@ Where:
+
+Combination of these two, it yiels to the following figure:
+
+<img src="documentation_data/mohr-coulomb-with-tension-cutoff-zones.svg" alt="Mohr-Coulomb with tension cutoff" title="Mohr-Coulomb with tension cutoff" width="800">
```
The symbol $`G`$ is used in the picture, but never introduced in the text.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/documentation_data/mohr-coulomb-with-tension-cutoff-zones.svg:
We've found a typo: "Regulaire" should be "Regular".

mnabideltares committed:
Applied patch bases on review

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/documentation_data/Mohr-Coulomb-with-tension_cutoff-formulations.pdf:
On page 5, the equations following the sentence "These equations can be rewritten as:" are never really used and don't help to better understand the matter, so they could be removed.

On the same page, the final equations for $`\sigma_1`$ and $`\sigma_3`$ can be expressed in vector notation, to follow more closely the corresponding suggested code.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_mohr_coulomb_with_tension_cutoff.cpp:
We would suggest to add a unit test, where the tensile strength equals the apex.

avdg81 reviewed (commented):
Thank you very much for processing the previous round of review suggestions. We have some more now that we have reviewed the code once again. We feel it is in a good shape, so most suggestions are truly minor. Hope this helps to make the code even better.

Please don't forget to look at the code smells reported by SonarQube and to slightly improve the code coverage. But overall, we are really satisfied with the work that you've done. Thank you.

mnabideltares committed:
Fixes based on review

mnabideltares committed:
Fixes

mnabideltares committed:
fix in readme

mnabideltares committed:
Added a unit test

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/13071-implement-mohr-coulomb-continuum

avdg81 committed:
Added (de)serialization of the tension cut-off

Other changes include:
- Added empty implementations of `save` and `load` of class `CoulombYieldSurface`, since any class that derives from `YieldSurface` must implement them, or else the code can't be compiled.
- Added a unit test to cover the `save` and `load` functionality of class `TensionCutoff`.
- Added functions to register and deregister yield surfaces with the serializer.
- Added scoped serializer registration for the yield surfaces (needed for the unit tests).
- Formatted the code base with `kp format`.
- Removed an empty statement.

avdg81 committed:
Added (de)serialization of the Coulomb yield surface

Other changes include:
- Added a unit test to cover the `save` and `load` functionality of class `CoulombYieldSurface`.
- Register and deregister the Coulomb yield surface with the serializer.

avdg81 committed:
Fixed a bad index error

mnabideltares committed:
Added unit tests for permutations

mnabideltares committed:
chanup

mnabideltares committed:
cleanup

mnabideltares committed:
Fixes based on review

mnabideltares committed:
fixes based on sonarcloud code smells

avdg81 committed:
Removed untested (de)serialization of class `MohrCoulombWithTensionCutOff`

The (de)serialization of this class will be addressed by a separate PR.

mnabideltares committed:
A fix, and added a unit test

avdg81 committed:
Improved the material property check

Don't repeat the checks. Also unit test out-of-bounds values.

avdg81 committed:
Avoid using intermediate local variables

avdg81 committed:
Reuse function that creates vector from diagonal matrix

avdg81 committed:
Get diagonal entries when matrix has at least as many columns as rows

Also, no longer check whether the off-diagonal entries are unequal to zero.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/13071-implement-mohr-coulomb-continuum

avdg81 committed:
Mostly stylistic changes

And fixed a typo.

avdg81 committed:
Stylistic changes only

avdg81 reviewed (approved):
Thank you very much for all the hard work that you have put into this PR. It was a lot, but it has become very clear and clean code, which is well-tested and has a high code coverage.
